File: LC086PartitionList.java
package leetcode;

/**
 * https://leetcode.com/problems/partition-list/
 */
public class LC086PartitionList {
    public class ListNode {
        int val;
        ListNode next;

        ListNode(int x) {
            val = x;
        }
    }

    private ListNode build(int[] array) {
        ListNode head = null;
        ListNode tail = null;
        for (int value : array) {
            if (head == null) {
                head = tail = new ListNode(value);
            } else {
                tail.next = new ListNode(value);
                tail = tail.next;
            }
        }
        tail.next = null;
        return head;
    }

    private void print(ListNode head) {
        while (head != null) {
            System.out.print("" + head.val + " ");
            head = head.next;
        }
        System.out.println();
    }

    public ListNode partition(ListNode head, int x) {
        if (head == null) {
            return null;
        }
        ListNode less = new ListNode(0);
        ListNode more = new ListNode(0);
        ListNode before = less;
        ListNode after = more;
        while (head != null) {
            if (head.val < x) {
                less.next = head;
                less = head;
            }
            if (head.val >= x) {
                more.next = head;
                more = head;
            }
            head = head.next;
        }
        less.next = after.next;
        more.next = null;
        return before.next;
    }

    public static void main(String[] arguments) {
        LC086PartitionList splitter = new LC086PartitionList();
        ListNode list = splitter.build(new int[]{1, 4, 3, 2, 5, 2});
        splitter.print(splitter.partition(list, 3));
    }
}

File: algorithms/code/leetcode/lc086_partition_list/lc086_partition_list.py
#!/usr/bin/env python3
# https://leetcode.com/problems/partition-list/

from __future__ import annotations

import unittest


class ListNode:
    def __init__(
        self, val: int = 0, following: 'ListNode' | None = None
    ) -> None:
        self.val = val
        self.next = following


class Solution:
    def partition(self, head: ListNode | None, x: int) -> ListNode | None:
        if head is None:
            return None
        less = ListNode(0)
        more = ListNode(0)
        before = less
        after = more
        while head is not None:
            if head.val < x:
                less.next = head
                less = head
            else:
                more.next = head
                more = head
            head = head.next
        less.next = after.next
        more.next = None
        return before.next


class TestCode(unittest.TestCase):
    def __build(self, array: list[int]) -> ListNode | None:
        head = None
        tail = None
        for value in array:
            if head is None:
                head = tail = ListNode(value)
            else:
                assert tail is not None
                tail.next = ListNode(value)
                tail = tail.next
        if tail is not None:
            tail.next = None
        return head

    def test_1_4_3_2_5_2__3(self) -> None:
        listed = self.__build([1, 4, 3, 2, 5, 2])
        expected = [1, 2, 2, 4, 3, 5]
        result = Solution().partition(listed, 3)
        for value in expected:
            assert result is not None
            assert value == result.val
            result = result.next

    def test_nothing(self) -> None:
        assert Solution().partition(None, 0) is None
